stats/patient_interceptor.go

```diff
@@ -0,0 +1,108 @@
+package stats
+
+import (
+	"errors"
+	"log"
+
+	"github.com/intervention-engine/fhir/models"
+)
+
+// PatientStatsCreateInterceptor intercepts any new patient resources added to the database
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsCreateInterceptor struct {
+	DataAccess StatsDataAccess
+}
+
+// Before is unused
+func (s *PatientStatsCreateInterceptor) Before(resource interface{}) {}
+
+// After increments population statistics after a patient resource is created.
+func (s *PatientStatsCreateInterceptor) After(resource interface{}) {
+	patient, ok := resource.(*models.Patient)
+
+	if ok {
+		err := s.DataAccess.AddPatientStat(patient)
+		if err != nil {
+			log.Printf("PatientStatsCreateInterceptor: After: %s\n", err.Error())
+		}
+	}
+}
+
+// OnError is unused
+func (s *PatientStatsCreateInterceptor) OnError(err error, resource interface{}) {}
+
+// PatientStatsUpdateInterceptor intercepts any updated patient resources
+// and updates the Synthetic Mass population statistics based on that patient's address.
+type PatientStatsUpdateInterceptor struct {
+	DataAccess StatsDataAccess
+	// The state of the patient before the database update, for comparison after the database update
+	patientBefore *models.Patient
+	// Tracks if the interceptor failed to cache the patient model before the update
+	cacheError error
```

Comment:
Golang maps are thread safe for multiple readers, no writers, or one writer no readers. Any other form of access would require synchronization.